graphql-native/src/main/java/io/ballerina/stdlib/graphql/engine/Engine.java

```diff
@@ -0,0 +1,251 @@
+import static io.ballerina.stdlib.graphql.utils.Constants.FIELD_SELECTIONS;
+import static io.ballerina.stdlib.graphql.utils.Constants.FIELD_TYPE;
+import static io.ballerina.stdlib.graphql.utils.Constants.FIELD_VALUE;
+import static io.ballerina.stdlib.graphql.utils.Constants.OPERATION_QUERY;
+import static io.ballerina.stdlib.graphql.utils.Utils.createError;
+import static io.ballerina.stdlib.graphql.utils.Utils.createErrorArray;
+import static io.ballerina.stdlib.graphql.utils.Utils.createFieldNotFoundError;
+import static io.ballerina.stdlib.graphql.utils.Utils.createResourceExecutionFailedError;
+
+/**
+ * This handles Ballerina GraphQL Engine.
+ */
+public class Engine {
+
+    /**
+     * Returns a stored resource value of a Ballerina service.
+     *
+     * @param listener - GraphQL listener to which the service is attached
+     * @param field    - The {@code Field} record from the Ballerina
+     * @return - Resource value
+     */
+    public static Object executeResource(BObject listener, BMap<BString, Object> field) {
+        BObject attachedService = (BObject) listener.getNativeData(Constants.NATIVE_SERVICE_OBJECT);
+        AttachedFunctionType[] attachedFunctions = attachedService.getType().getAttachedFunctions();
+        String fieldName = field.get(FIELD_NAME).toString();
+        for (AttachedFunctionType attachedFunction : attachedFunctions) {
+            if (attachedFunction.getName().equals(fieldName)) {
+                return Wrapper.invokeResource(attachedFunction, null);
+            }
+        }
+        return createResourceExecutionFailedError(fieldName, OPERATION_QUERY);
+    }
+
+    /**
+     * Returns an array of field names for the attached service for a Ballerina GraphQL listener.
+     *
+     * @return - A {@code BArray} consisting field names for the given service
+     */
+    public static BArray getFieldNames(BObject service) {
+        AttachedFunctionType[] attachedFunctions = service.getType().getAttachedFunctions();
+        String[] result = new String[attachedFunctions.length];
+        for (int i = 0; i < attachedFunctions.length; i++) {
+            result[i] = attachedFunctions[i].getName();
+        }
+        return ValueCreator.createArrayValue(StringUtils.fromStringArray(result));
+    }
+
+    /**
+     * Validates a given field in a GraphQL document with the schema.
+     *
+     * @param listener  - Ballerina GraphQL listener object
+     * @param field     - The GraphQL operation field which needs to be validated
+     * @param operation - The type of the operation
+     * @return - A {@code BArray} consisting field names for the given service
+     */
+    public static Object validateField(BObject listener, BMap<BString, Object> field, BString operation) {
+        BObject service = (BObject) listener.getNativeData(Constants.NATIVE_SERVICE_OBJECT);
+        return validateFieldWithService(field, service, operation);
+    }
```

Comment:
I think we need to do other way around instead of passing the built Document tree to an extern function to validate, what we need to do is have some extern functions to help us build a tree for schema definition in Ballerina. Then the validation logic of Document can be written in Ballerina itself. 

I think that is what we agreed during the meeting we had with Sanjiva too if I am not mistaken.

Reply:
Yes, as I mentioned in the code review, we should build the graph when a service is attached. This is done just for now since we don't have the APIs to access that service information just yet. Once we get that information, will change to the mentioned way.

Reply:
Great! But on second thoughts that sounds a bit strange too. Because we have the tree-like service definition written in Ballerina, then we go from there to java, then we go from java to Ballerina again to build some sort of tree value from the service definition. 

Anyway, let's have a chat on this.

Reply:
Yes, that seems quite messy. Will try to come up with a better solution.